plugins/telink-dfu/fu-telink-dfu-hid-device.c

```diff
@@ -0,0 +1,409 @@
+		ota_data_len = FU_TELINK_DFU_HID_DEVICE_OTA_END_LEN;
+		break;
+	default:
+		/* OTA data */
+		ota_data_len = FU_TELINK_DFU_HID_DEVICE_OTA_DATA_LEN;
+		break;
+	}
+
+	fu_struct_telink_dfu_hid_pkt_payload_set_ota_cmd(payload, cmd);
+	if (buf != NULL)
+		fu_struct_telink_dfu_hid_pkt_payload_set_ota_data(payload, buf, bufsz, NULL);
+	/* exclude the ota_cmd field */
+	fu_struct_telink_dfu_hid_pkt_payload_set_crc(
+	    payload,
+	    ~fu_crc16(FU_CRC_KIND_B16_USB, payload->data, payload->len - 2));
+	fu_struct_telink_dfu_hid_pkt_set_ota_data_len(pkt, ota_data_len);
+	fu_struct_telink_dfu_hid_pkt_set_payload(pkt, payload, error);
+	return pkt;
+}
+
+static gboolean
+fu_telink_dfu_hid_device_write(FuTelinkDfuHidDevice *self, guint8 *buf, gsize bufsz, GError **error)
+{
+	FuHidDevice *hid_dev = FU_HID_DEVICE(self);
+	FuHidDeviceFlags set_report_flag = FU_HID_DEVICE_FLAG_NONE;
+
+	if (self->hid_tool_ver >= HID_TOOL_VER(5, 2)) {
+		set_report_flag = FU_HID_DEVICE_FLAG_USE_INTERRUPT_TRANSFER;
+	}
+
+	if (!fu_hid_device_set_report(hid_dev,
+				      FU_TELINK_DFU_HID_DEVICE_REPORT_ID,
+				      buf,
+				      FU_TELINK_DFU_HID_DEVICE_OTA_LENGTH,
+				      FU_TELINK_DFU_HID_DEVICE_REPORT_TIMEOUT,
+				      set_report_flag,
+				      error)) {
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_telink_dfu_hid_device_write_blocks(FuTelinkDfuHidDevice *self,
+				      FuChunkArray *chunks,
+				      FuProgress *progress,
+				      GError **error)
+{
+	FuStructTelinkDfuHidLongPkt *long_pkt = fu_struct_telink_dfu_hid_long_pkt_new();
+	guint payload_index;
+	guint i;
+
+	/* progress */
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_set_steps(progress, fu_chunk_array_length(chunks));
+	for (i = 0; i < fu_chunk_array_length(chunks); i++) {
+		g_autoptr(FuChunk) chk = NULL;
+		g_autoptr(FuStructTelinkDfuHidPkt) pkt = NULL;
+
+		/* send chunk */
+		chk = fu_chunk_array_index(chunks, i, error);
+		if (chk == NULL)
+			return FALSE;
+		pkt = fu_telink_dfu_hid_device_create_packet((guint16)i,
+							     fu_chunk_get_data(chk),
+							     fu_chunk_get_data_sz(chk),
+							     error);
+		if (self->hid_tool_ver >= HID_TOOL_VER(5, 2)) {
+			FuStructTelinkDfuHidPktPayload *payload =
+			    fu_struct_telink_dfu_hid_pkt_get_payload(pkt);
+			/* TODO: find a better method to declare the structure array */
+			payload_index = i % 3;
+			fu_struct_telink_dfu_hid_long_pkt_set_ota_data_len(
+			    long_pkt,
+			    FU_TELINK_DFU_HID_DEVICE_OTA_DATA_LEN * (payload_index + 1));
+			if (payload_index == 0) {
+				FuStructTelinkDfuHidPktPayload *default_payload =
+				    fu_struct_telink_dfu_hid_pkt_payload_new();
+				fu_struct_telink_dfu_hid_long_pkt_set_payload_1(long_pkt,
+										payload,
+										error);
+				fu_struct_telink_dfu_hid_long_pkt_set_payload_2(long_pkt,
+										default_payload,
+										error);
+				fu_struct_telink_dfu_hid_long_pkt_set_payload_3(long_pkt,
+										default_payload,
+										error);
```

Comment:
you need to check the success of each of these, and if fail, return FALSE;

Reply:
resoved in commit 0fb74d587e8aedf6a6bccabff38623ea034d224d